pkg/cryptoauthlib/contrib/atca.c

```diff
@@ -0,0 +1,204 @@
+    uint8_t retries = cfg->rx_retries;
+    int ret = -1;
+    uint8_t length_package = 0;
+    uint8_t bytes_to_read;
+
+    /* read first byte (size of output data) and store it in variable length_package
+    to check if output will fit into rxdata */
+    i2c_acquire(cfg->atcai2c.bus);
+    while (retries-- > 0 && ret != 0)
+    {
+        ret = i2c_read_byte(cfg->atcai2c.bus, (cfg->atcai2c.slave_address >> 1), 
+        &length_package, 0);
+    }
+    i2c_release(cfg->atcai2c.bus);
+
+    if (ret != 0)
+    {
+        return ATCA_RX_TIMEOUT;
+    }
+
+    bytes_to_read = length_package-1;
+
+    if (bytes_to_read > *rxlength)
+    {
+        return ATCA_SMALL_BUFFER;
+    }
+
+    /* CRC function calculates value of the whole output package, so to get a correct result we need to include the length of the package we got before into rxdata as first byte. */
+    rxdata[0] = length_package;
+
+    /* reset ret and retries to read the rest of the output */
+    ret = -1;
+    retries = iface->mIfaceCFG->rx_retries;
+
+    /* read rest of output and insert into rxdata array after first byte */
+    i2c_acquire(cfg->atcai2c.bus);
+    while (retries-- > 0 && ret != 0)
+    {
+        ret = i2c_read_bytes(cfg->atcai2c.bus, (cfg->atcai2c.slave_address >> 1), (rxdata + 1), bytes_to_read, 0);
+    }
+    i2c_release(cfg->atcai2c.bus);
+
+    if (ret != 0)
+    {
+        return ATCA_RX_TIMEOUT;
+    }
+
+    *rxlength = length_package;
+
+    return ATCA_SUCCESS;
+}
+
+ATCA_STATUS hal_i2c_wake(ATCAIface iface)
+{
+    ATCAIfaceCfg *cfg = atgetifacecfg(iface);
+    /* ATCA_PARAM_I2C needs to be woken up by holding the sda pin low for some time and then reinitializing it */
+    /* SDA as GPIO, Output to manually set it to low */
+    if(gpio_init(ATCA_GPIO_WAKE, GPIO_OUT) == -1)
+    {
+        return ATCA_GEN_FAIL;
+    }
+    gpio_clear(ATCA_GPIO_WAKE);
+    /* wait 30 us (t(WLO)) */
+    xtimer_usleep(30);
+    
+    /* reinitialize i2c-ATCA_PARAM_I2C */
+    i2c_init(cfg->atcai2c.bus);
+    /* wait 1500 us (t(WHI)) */
+    xtimer_usleep(1500);
+    return ATCA_SUCCESS;
+}
+
+ATCA_STATUS hal_i2c_idle(ATCAIface iface)
+{
+    ATCAIfaceCfg *cfg = atgetifacecfg(iface);
+
+    /* idle state = write byte to register adr. 0x02 */
+    uint8_t idle = ATCA_IDLE_ADR;
+
+    i2c_acquire(cfg->atcai2c.bus);
+    i2c_write_byte(cfg->atcai2c.bus, (cfg->atcai2c.slave_address >> 1), idle, 0);
+    i2c_release(cfg->atcai2c.bus);
+    return ATCA_SUCCESS;
+}
+
+ATCA_STATUS hal_i2c_sleep(ATCAIface iface)
+{
+    ATCAIfaceCfg *cfg = atgetifacecfg(iface);
+
+    /* sleep state = write byte to register adr. 0x01 */
+    uint8_t sleep = ATCA_SLEEP_ADR;
+
+    i2c_acquire(cfg->atcai2c.bus);
+    i2c_write_byte(cfg->atcai2c.bus, (cfg->atcai2c.slave_address >> 1), sleep, 0);
+    i2c_release(cfg->atcai2c.bus);
+    return ATCA_SUCCESS;
+}
+
+ATCA_STATUS hal_i2c_release(void *hal_data)
+{
+    ATCAIfaceCfg* cfg = (ATCAIfaceCfg*)hal_data;
+    i2c_release(cfg->atcai2c.bus);
```

Comment:
When running the library's tests this function is still needed (at least they don't run if I leave this unimplemented)

Reply:
This is weird and might point to an other issue. Eventually, the library expects this function to return success so this should stay in. But if a "missing" `i2c_release` blocks here because it waits for the mutex to unlock, it might indicate a missing release somewhere else. Can you check which acquire is the last before blocking?